lib/Data/Time/Clock/POSIX.hs

```diff
@@ -1,84 +1,127 @@
 -- | POSIX time, if you need to deal with timestamps and the like.
 -- Most people won't need this module.
 module Data.Time.Clock.POSIX
-(
-    posixDayLength,POSIXTime,posixSecondsToUTCTime,utcTimeToPOSIXSeconds,getPOSIXTime,getCurrentTime
-) where
+    ( -- *  POSIX timestamps
+      POSIXTime(..)
+    , normalizePosix
+    , posixToUTCTime
+    , getPOSIXTime
+    , unixEpochDay
+    , posixDayLength
+    , getCurrentTime
+    , posixSecondsToUTCTime
+    , utcTimeToPOSIXSeconds
+    ) where
 
 import Data.Time.Clock.UTC
+import Data.Time.Clock.Scale (picosecondsToDiffTime)
 import Data.Time.Calendar.Days
-import Data.Fixed
-import Control.Monad
+import Data.Int (Int64)
+import Data.Fixed (divMod')
+import Control.DeepSeq
 
 #include "HsTimeConfig.h"
 
 #ifdef mingw32_HOST_OS
-import Data.Word    ( Word64)
+import Data.Word (Word64)
 import System.Win32.Time
 #elif HAVE_CLOCK_GETTIME
 import Data.Time.Clock.CTimespec
-import Foreign.C.Types (CTime(..))
+import Foreign.C.Types (CTime(..), CLong(..))
 #else
 import Data.Time.Clock.CTimeval
+import Foreign.C.Types (CLong(..))
 #endif
 
--- | 86400 nominal seconds in every day
-posixDayLength :: NominalDiffTime
-posixDayLength = 86400
+--------------------------------------------------------------------------------
 
 -- | POSIX time is the nominal time since 1970-01-01 00:00 UTC
 --
--- To convert from a 'Foreign.C.CTime' or 'System.Posix.EpochTime', use 'realToFrac'.
---
-type POSIXTime = NominalDiffTime
+data POSIXTime = POSIXTime
+    { ptSeconds ::     {-# UNPACK #-} !Int64
+    , ptNanoSeconds :: {-# UNPACK #-} !Int64
```

Comment:
Might as well make `ptNanoSeconds` `Word32`, given that it gets converted anyway.